src/rust-crypto/libolm_migration.ts

```diff
@@ -380,3 +382,91 @@ async function getAndDecryptCachedSecretKey(
+ * By default, if the legacy session didn't have the private MSK, the migrated session will revert to unverified,
+ * even if the user has verified the session in the past.
+ * This only occurs if the private MSK was not cached in the crypto store (USK and SSK private keys won't help
+ * to establish trust: the trust is rooted in the MSK).
+ * Rust crypto will only consider the current session as trusted if we import the private MSK itself.
+ *
+ * We could prompt the user to verify the session again, but it's probably better to just mark the user identity
+ * as locally verified if it was before.
+ *
+ * See https://github.com/element-hq/element-web/issues/27079
+ *
+ * @param args - Argument object.
+ */
+export async function migrateLegacyLocalTrustIfNeeded(args: {
+    /** The legacy crypto store that is migrated. */
+    legacyCryptoStore: CryptoStore;
+    /** The migrated rust crypto stack. */
+    rustCrypto: RustCrypto;
+    /** The logger to use */
+    logger: Logger;
+}): Promise<void> {
+    const { legacyCryptoStore, rustCrypto, logger } = args;
+    // Now get the cross-signing identity from rust.
+    // If this is null that means that there are no cross-signing keys published server side.
+    const rustOwnIdentity = await rustCrypto.getOwnIdentity();
+    if (!rustOwnIdentity) {
+        // There are no cross-signing keys published server side, so nothing to do here.
+        return;
+    }
+    if (rustOwnIdentity.isVerified()) {
+        // The rust session already trusts the keys, so again, nothing to do.
+        return;
+    }
+
+    // This would be null is the user never verified their identity in the legacy session.
+    const legacyLocallyTrustedMSK = await getLegacyTrustedPublicMasterKeyBase64(legacyCryptoStore);
+    if (!legacyLocallyTrustedMSK) {
+        // Nothing to do, the legacy session was not verified
+        return;
+    }
+
+    const mskInfo: CrossSigningKeyInfo = JSON.parse(rustOwnIdentity.masterKey);
+    if (!mskInfo.keys || Object.keys(mskInfo.keys).length === 0) {
+        // This should not happen, but let's be safe
+        logger.error("Post Migration | Unexpected error: no master key in the rust session.");
+        return;
+    }
+    const rustSeenMSK = Object.values(mskInfo.keys)[0];
+
+    if (rustSeenMSK && rustSeenMSK == legacyLocallyTrustedMSK) {
+        logger.info(`Post Migration: Migrating legacy trusted MSK: ${legacyLocallyTrustedMSK} to locally verified.`);
+        // Let's mark the user identity as locally verified as part of the migration.
+        await rustOwnIdentity!.verify();
+        // This will also sign the MSK with the current device key, and return the request to be sent to the server.
+        // We ignored it in our case. As the local trust was migrated and not actually done by the user it looks
+        // sane to not publish the signature. Anyhow, these device's signatures are not considered by the rust-sdk to
+        // establish the trust of the user identity.
```

Comment:
The part about rust-sdk not using the signatures doesn't seem very compelling: couldn't you make the same argument for other places we call `.verify`?

Reply:
We need to clarify when we should add device signature on the MSK, tracked here https://github.com/element-hq/crypto-internal/issues/190